### PR TITLE
Update NeMo/Megatron

### PR DESCRIPTION
**Pull request by sichu2023**
## Summary
* avoid checkpoint directory name clashes
* `IOMixin` only serializes non-default values
* update ESM-2 tokenizer serialization test
* turn off `ckpt_async_save` in stop-and-go test
* switch to `trainer.should_stop` instead of raising exception
* identify commit `19766a217` which impact training resumption

## Details
Megatron now checks whether a non-empty checkpoint directory exists before overwriting ([commit](https://github.com/NVIDIA/Megatron-LM/commit/8f982449847bac1e367b7609af1269b13bd29b48#diff-67e985c7c0f975e4463533f53f578597310392a21629929fa2b5c89af13b2112R115)). Our unittests have a small `validate_every_n_steps` which leads to identical monitored metric, and thus the checkpoint directory name.

NeMo updated `IOMixin.io_dump` to include `yaml_attrs`. Also, `IOMixin` only serializes non-default values due to constant version conflicts in Megatron config objects.

## Notes
Apparently there are unittest leakage but only on a few test functions listed below.

## Pytest errors
On NeMo/Megatron TOT
```bash
FAILED sub-packages/bionemo-esm2/tests/bionemo/esm2/model/test_stop_and_go.py::TestESM2StopAndGo::test_stop_and_go_consistency[ValidOutputCallback] - AssertionError: Tensor-likes are not close!
FAILED sub-packages/bionemo-esm2/tests/bionemo/esm2/model/test_stop_and_go.py::TestESM2StopAndGo::test_stop_and_go_consistency[ValidLossCallback] - AssertionError: Scalars are not close!
```

Error from test leakage
```bash
FAILED sub-packages/bionemo-esm2/tests/bionemo/esm2/model/finetune/test_finetune.py::test_esm2_finetune_token_classifier[False] - TypeError: unsupported operand type(s) for /: 'PosixPath' and 'int'
FAILED sub-packages/bionemo-esm2/tests/bionemo/esm2/model/finetune/test_finetune.py::test_esm2_finetune_regressor[False] - TypeError: unsupported operand type(s) for /: 'PosixPath' and 'int'
FAILED sub-packages/bionemo-example_model/tests/bionemo/example_model/test_lightning_basic.py::test_train_mnist_litautoencoder_with_megatron_strategy_single_gpu[32] - torch.distributed.checkpoint.api.CheckpointException: CheckpointException ranks:dict_keys([0])
FAILED sub-packages/bionemo-example_model/tests/bionemo/example_model/test_lightning_basic.py::test_train_mnist_litautoencoder_with_megatron_strategy_single_gpu[bf16-mixed] - torch.distributed.checkpoint.api.CheckpointException: CheckpointException ranks:dict_keys([0])
FAILED sub-packages/bionemo-testing/tests/bionemo/testing/data/test_load.py::test_default_pbss_client - botocore.exceptions.ConfigParseError: Unable to parse config file: /home/bionemo/.aws/config
```

### CODE REVIEW

**Comment by sichu2023** [comment]
Related to https://github.com/NVIDIA/bionemo-framework/pull/253

**Comment by sichu2023** [comment]
Observe warning regarding `LRScheduler`
```bash
------------------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------------------
Trainer already configured with model summary callbacks: [<class 'pytorch_lightning.callbacks.rich_model_summary.RichModelSummary'>]. Skipping setting a default `ModelSummary` callback.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
[NeMo W 2024-10-31 15:26:44 nemo_logger:123] No version folders would be created under the log folder as 'resume_if_exists' is enabled.
[NeMo W 2024-10-31 15:26:44 nemo_logger:173] "update_logger_directory" is True. Overwriting tensorboard logger "save_dir" to /tmp/pytest-of-bionemo/pytest-13/test_esm2_finetune_token_class0/pretrain/../../tmp/pytest-of-bionemo/pytest-13/test_esm2_finetune_token_class0/pretrain
[NeMo W 2024-10-31 15:26:44 nemo_logger:189] The Trainer already contains a ModelCheckpoint callback. This will be overwritten.
[NeMo W 2024-10-31 15:26:44 resume:215] There were no checkpoints found in checkpoint_dir or no checkpoint folder at checkpoint_dir :/tmp/pytest-of-bionemo/pytest-13/test_esm2_finetune_token_class0/pretrain/test_experiment/checkpoints. Training from scratch.
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1]
[NeMo W 2024-10-31 15:26:44 megatron_strategy:310] Could not copy Trainer's 'max_steps' to LR scheduler's 'max_steps'. If you are not using an LR scheduler, this warning can safely be ignored.
```

**Comment by sichu2023** [comment]
Related to https://github.com/NVIDIA/TransformerEngine/pull/1130

**Comment by sichu2023** [comment]
@jstjohn Looking into `sub-packages/bionemo-llm/tests/bionemo/llm/utils/test_iomixin_utils.py::TestIOMixin::test_dataclass_out_of_sync`, it seems that default_factory class attribute "a" and inherited "c" are no longer accessible by `get_hparams`.

```
(Pdb) v1.a, v1.b, v1.c
(4, 3, 3)
(Pdb) v1.get_hparams()
{'b': 7} 
```

File: scripts/protein/esm2/test_esm2_pretrain.py [hunk 1]
```diff
@@ -62,7 +61,6 @@ def dummy_protein_dataset(tmp_path):
     return db_file
 
 
-@pytest.mark.skip("duplicate unittest")
 @pytest.fixture
 def dummy_parquet_train_val_inputs(tmp_path):
```
Comment (farhadrgh) [inline scripts/protein/esm2/test_esm2_pretrain.py:65]: now that you're at it you may want to consider importing these fixtures from bionemo.testing

**Review by farhadrgh — APPROVED** [review]

File: Dockerfile [hunk 2]
```diff
@@ -17,7 +17,7 @@ RUN git clone https://github.com/NVIDIA/apex.git && \
   --config-settings "--build-option=--cpp_ext --cuda_ext --fast_layer_norm --distributed_adam --deprecated_fused_adam --group_norm"
 
 # Transformer Engine pre-1.7.0. 1.7 standardizes the meaning of bits in the attention mask to match
-ARG TE_COMMIT=7d576ed25266a17a7b651f2c12e8498f67e0baea
+ARG TE_COMMIT=c27ee60ec746210bcea4ec33958dbbff06706506
```
Comment (pstjohn) [inline Dockerfile:20]: Attempting to split this as a separate PR in #399

File: sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py [hunk 3]
```diff
@@ -217,6 +217,14 @@ def make_callbacks() -> Dict[Type[pl.Callback], pl.Callback]:
                 testing_callbacks.ValidInputCallback: testing_callbacks.ValidInputCallback(),
                 testing_callbacks.ValidOutputCallback: testing_callbacks.ValidOutputCallback(),
                 testing_callbacks.ValidLossCallback: testing_callbacks.ValidLossCallback(),
+                nl_callbacks.ModelCheckpoint: nl_callbacks.ModelCheckpoint(
```
Comment (pstjohn) [inline sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py:220]: moving this here (and instantiating it once for both stop and resume branches) is what's causing that `deepcopy` error.

File: sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py [hunk 4]
```diff
@@ -234,10 +234,11 @@ def make_callbacks() -> Dict[Type[pl.Callback], pl.Callback]:
                 testing_callbacks.RaiseAfterMetadataCallback: testing_callbacks.RaiseAfterMetadataCallback(),
                 nl_callbacks.ModelCheckpoint: nl_callbacks.ModelCheckpoint(
                     save_last=True,
-                    monitor="reduced_train_loss",
+                    monitor="val_loss",
                     save_top_k=2,
                     every_n_train_steps=cls.val_check_interval,
                     always_save_context=True,
+                    filename="{epoch}-{step}-{val_loss:.2f}",
                 ),
```
Comment (pstjohn) [inline sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py:242]: Even with these changes, I still get the `Last checkpoint is unfinished and cannot be used to resume the training` error

Comment (sichu2023) [reply sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py:243]: This commit should now fix it.
https://github.com/NVIDIA/bionemo-framework/pull/302/commits/f2f6d60694c6443731b16f7356f07c89124a4ee1

**Comment by sichu2023** [comment]
Dependent on https://github.com/NVIDIA/bionemo-framework/pull/414

**Comment by sichu2023** [comment]
NeMo quick patch to fix stop and go. Can remove xfail and merge after NeMo's PR.
https://github.com/NVIDIA/NeMo/pull/11029

**Comment by sichu2023** [comment]
Here is the NeMo long term fix.
https://github.com/Lightning-AI/pytorch-lightning/pull/20379

**Comment by sichu2023** [comment]
Issue tracking: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/issues/888

File: sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py [hunk 5]
```diff
@@ -335,12 +336,27 @@ def test_stop_and_go_consistency(self, callback_type):
         continuous_callback = get_callback(self.callbacks, Mode.CONTINUOUS, callback_type)
         assert interrupted_callback.data, f"No data found for {callback_type}"
 
-        if callback_type == testing_callbacks.TrainOutputCallback:
-            atol = 1e-3
+        if callback_type in {testing_callbacks.TrainOutputCallback, testing_callbacks.ValidOutputCallback}:
+            atol, rtol = 2e-2, 0  # no relative tolerance when logits can be zero
```
Comment (pstjohn) [inline sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py:340]: @jstjohn, @sichu2023; just flagging this line as a possible problem that geneformer outputs are so different between runs. We can set `atol, rtol = 1E-3, 1E-4` for ESM-2; and we used to be able to set that for geneformer before bumping versions
